include/ginkgo/core/distributed/collective_communicator.hpp

```diff
@@ -0,0 +1,117 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_COLLECTIVE_COMMUNICATOR_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_COLLECTIVE_COMMUNICATOR_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/index_map_fwd.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace mpi {
+
+/**
+ * Interface for an collective communicator.
+ *
+ * A collective communicator only provides routines for collective
+ * communications. At the moment this is restricted to the variable all-to-all.
+ */
+class CollectiveCommunicator {
+public:
+    virtual ~CollectiveCommunicator() = default;
+
+    explicit CollectiveCommunicator(communicator base = MPI_COMM_NULL);
+
+    [[nodiscard]] const communicator& get_base_communicator() const;
+
+    /**
+     * Non-blocking all-to-all communication.
+     *
+     * The send_buffer must have allocated at least get_send_size number of
+     * elements, and the recv_buffer must have allocated at least get_recv_size
+     * number of elements.
+     *
+     * @tparam SendType  the type of the elements to send
+     * @tparam RecvType  the type of the elements to receive
+     * @param exec  the executor for the communication
+     * @param send_buffer  the send buffer
+     * @param recv_buffer  the receive buffer
+     * @return  a request handle
+     */
+    template <typename SendType, typename RecvType>
+    request i_all_to_all_v(std::shared_ptr<const Executor> exec,
+                           const SendType* send_buffer,
+                           RecvType* recv_buffer) const
+    {
+        return this->i_all_to_all_v(
+            std::move(exec), send_buffer, type_impl<SendType>::get_type(),
+            recv_buffer, type_impl<RecvType>::get_type());
+    }
+
+    /**
+     * @copydoc i_all_to_all_v(std::shared_ptr<const Executor>, const SendType*
+     *          send_buffer, RecvType* recv_buffer)
+     */
+    virtual request i_all_to_all_v(std::shared_ptr<const Executor> exec,
+                                   const void* send_buffer,
+                                   MPI_Datatype send_type, void* recv_buffer,
+                                   MPI_Datatype recv_type) const = 0;
```

Comment:
That is a good point! I will update this.

Reply:
Thinking about it again, it doesn't make much sense to make this private, since we need access to the `MPI_Datatype` for communicating vectors with multiple RHS.

Reply:
Not make it private, but make it non-virtual backed by a private implementation

Reply:
Right, that makes more sense.